server/src/main/java/io/crate/execution/dml/ArrayIndexer.java

```diff
@@ -34,10 +35,32 @@
 
 import io.crate.metadata.ColumnIdent;
 import io.crate.metadata.Reference;
+import io.crate.metadata.table.TableInfo;
+import io.crate.types.DataType;
 
 public class ArrayIndexer<T> implements ValueIndexer<List<T>> {
 
-    public static String toArrayLengthFieldName(Reference arrayRef) {
+    public static String toArrayLengthFieldName(Reference arrayRef, Function<ColumnIdent, Reference> getRef) {
+        // we want arrayRef's valueType not readType - i.e.:: the valueType of obj_array['int_col'] is 'int' while
+        // its readType is 'array(int)' - array length is indexed for 'obj_array' not for 'obj_array['int_col']
+        DataType<?> valueType = Optional
+            .ofNullable(getRef.apply(arrayRef.column())) // null if the arrayRef is dynamically added
+            .orElse(arrayRef).valueType();
+
+        // if the arrayRef is a ReadReference
+        if (!arrayRef.valueType().equals(valueType)) {
+            // 'o_array' is indexed but `o_array['xs']` is not indexed. We can use the fact that
+            // for every element of 'o_array' there exists its sub-column 'xs' such that
```

Comment:
2 comments:
Looks like 'o_array' and `o_array['xs']` are referring to examples from test which is a bit confusing. I think we need more generic description, like top level arrays have their length indexed but arrays which are object sub-columns don't. BTW is that true? Checking ArrayIndexer, looks like we index any array and also correctly add leaf name?

Reply:
I cherry-picked first commit and test only changes from the second commit.

Then I run `test_array_length_on_object_arrays_subcolumn`

Query query = convert("array_length(o_array['xs'], 1) >= 1");
assertThat(query).hasToString("_array_length_o_array:[1 TO 2147483647]");

now fails, query is :_array_length_xs:[1 TO 2147483647] - but it actually looks okay? We do utilize index and it comes to corect column (xs). OID is not used because of the test setup to be human-readable but in prod it will be leaf oid (or test can be adjusted). 

So in general, it looks to me that we need only tests from the second commit but the code change?

Reply:
Sorry I am having a hard time describing this. 

Using below as the baseline:
```
cr> create table t (o_array array(object as (xs int[])));                                                                                                             
CREATE OK, 1 row affected  (1.698 sec)

cr> insert into t values ([{xs=[1,2,3]}, {}]);                                                                                                                        
INSERT OK, 1 row affected  (0.190 sec)

cr> select o_array, array_length(o_array, 1) from t;                                                                                                                  
+-------------------------+--------------------------+
| o_array                 | array_length(o_array, 1) |
+-------------------------+--------------------------+
| [{"xs": [1, 2, 3]}, {}] |                        2 |
+-------------------------+--------------------------+
SELECT 1 row in set (0.003 sec)

cr> select o_array['xs'], array_length(o_array['xs'], 1) from t;                                                                                                      
+-------------------+--------------------------------+
| o_array['xs']     | array_length(o_array['xs'], 1) |
+-------------------+--------------------------------+
| [[1, 2, 3], null] |                              2 |
+-------------------+--------------------------------+
SELECT 1 row in set (0.002 sec)

cr> select o_array['xs'][1], array_length(o_array['xs'][1], 1) from t;                                                                                                
+------------------+-----------------------------------+
| o_array[1]['xs'] | array_length(o_array[1]['xs'], 1) |
+------------------+-----------------------------------+
| [1, 2, 3]        |                                 3 |
+------------------+-----------------------------------+
SELECT 1 row in set (0.004 sec)

cr> select o_array['xs'][2], array_length(o_array['xs'][2], 1) from t;                                                                                                
+------------------+-----------------------------------+
| o_array[2]['xs'] | array_length(o_array[2]['xs'], 1) |
+------------------+-----------------------------------+
|             NULL |                              NULL |
+------------------+-----------------------------------+
SELECT 1 row in set (0.003 sec)
```

The single row inserted created the following indexes:
- o_array
  - xs (per o_array element)
  - _array_length_xs (per o_array element)
- _array_length_o_array

Then:
`array_length(o_array, 1)` --> `_array_length_o_array`
`array_length(o_array['xs'], 1)` -x-> `_array_length_xs` (do not be confused with `_array_length_xs[1]`)
`array_length(o_array['xs'], 1)` == `array_length(o_array, 1)` --> `_array_length_o_array`

Reply:
BTW, 
```
- IntField(xs, 1)
- IntField(xs, 2)
- IntField(xs, 3)
- IntField(_array_length_xs, 3)
- IntField(xs, 4)
- IntField(xs, 5)
- IntField(_array_length_xs, 2)
- IntField(_array_length_o_array, 2)
```
this leads to a bug for arrays in tables before `5.9` because we count the number of `xs` using `NumTermsPerDocQuery`:
```
cr> select array_length(o_array['xs'], 1) from t where array_length(o_array['xs'], 1) >= 5;                                                                           
+--------------------------------+
| array_length(o_array['xs'], 1) |
+--------------------------------+
|                              2 |
+--------------------------------+

cr> select array_length(o_array['xs'], 1) from t where array_length(o_array['xs'], 1) > 4;                                                                            
+--------------------------------+
| array_length(o_array['xs'], 1) |
+--------------------------------+
|                              2 |
+--------------------------------+
```

Reply:
I think https://github.com/crate/crate/issues/16448 should be resolvd first so putting this on hold. @BaurzhanSakhariev

Reply:
>array_length(o_array, 1) --> _array_length_o_array

So basically, for top level arrays we can utilize its own indexed field in `WHERE array_len(top_level_array)... `statements?

>array_length(o_array['xs'], 1)...  it is neither _array_length_xs[1] nor _array_length_xs[2]
You mean this field is useless for using index correctly in `WHERE array_length(o_array['xs'], 1) `

// We can use the fact that
// for every element of 'o_array' there exists its sub-column 'xs' such that
// array_length(o_array, 1) = array_length(o_array['xs'], 1)

I still don't get it. 
What if we replace initial insert statement to 
`insert into t values ([{xs=[1,2,3]}, {xs=[4,5,6,7]}]); `
one sub-array has length 3 and another 4, both are not equal to array_length(o_array, 1) which is 2?

Reply:
If you `select o_array['xs'] from t` you will get `[[1,2,3], [4,5,6,7]]` where `[[1,2,3], [4,5,6,7]][1]` is `[1,2,3]` and `[[1,2,3], [4,5,6,7]][2]` is `[4,5,6,7]` hence 2 elements.

Reply:
I got confused with the read reference thing, now I got it after checking javadocs of the `getReadType`. 

I forgot that array type always goes through the `SourceLookup` and we extract those nested arrays and get combined result.

Ie despite on seeing `select o_array['xs']` result I had in my mind only granular access, o_array[0]['xs'] -> array, o_array[1]['xs'] => array.